P4RT_App_DB_Schema.md

```diff
@@ -0,0 +1,396 @@
+* Ternary is matching on a masked value. This is only available for P4 match kinds of `ternary`.
+* LPM is matching on a value with a given prefix length. This is only available for P4 match kinds `lpm`.
+
+**Value Formatting**
+
+Values are formatted according to their type (e.g. IPv4 addresses as `10.0.0.1`). The list of formats and their description are detailed below.
+
+```
+// Describes the format of a value which has a bit width `num_bits` (unless for
+
+// the STRING format, which has no limit in the number of characters).
+enum Format {
+ // Hex string, e.g. 0x0a8b. All lowercase, and always of length
+ // ceil(num_bits/4)+2 (1 character for every 4 bits, zero-padded to be
+ // divisible by 4, and 2 characters for the '0x' prefix).
+ HEX_STRING = 0;
+ // MAC address, e.g. 00:11:ab:cd:ef:22. All lowercase, and always 17
+ // characters long.
+ MAC = 1;
+ // IPv4 address, e.g. 10.0.0.2.
+ IPV4 = 2;
+ // IPv6 address, e.g. fe80::21a:11ff:fe17:5f80. All lowercase, formatted
+ // according to RFC5952. This can be used for any num_bits of 128 or less. If
+ // num_bits is less than 128, then by convention only the upper num_bits bits
+ // can be set.
+ IPV6 = 3;
+ // String format, only printable characters.
+ STRING = 4;
+}
+```
+
+
+**Port names:**
+
+Port names refer to the ports used in SONiC and use `Format::STRING`, example “Ethernet0”.
+
+
+## Schema definition
+
+
+### APP_DB
+
+
+#### P4RT Table
+
+All data in the Application Database (APPL DB) published by the P4RT application resides in  the P4RT table. Logically, the P4RT table is divided into **subtables**:
+
+* **DEFINITION** tables describe configuration of the logical tables.
+* **ACL_*** tables contain the runtime entries for ACL tables.
+* **FIXED_*** tables contain the runtime entries fixed SAI pipeline tables.
+
+##### 1. DEFINITION Tables
+
+Definition tables describe the configuration of logical tables. Each key represents a separate logical table. Definition tables provide more details about the table, like which stage of the pipeline they fit into, one or more match/action parameters and values. This helps the P4Orch layer to interpret the actual table entries (when they arrive later as part of the P4 Write request processing) and program them accordingly into the ASIC_DB. Currently, fixed tables do not need a definition and only ACL tables have corresponding DEFINITION tables.
+
+**DEFINITION Table Keys**
+
+The keys of a DEFINITION table represent the dynamic table name or the pipeline stage of the table that it defines.
+
+; defines the P4RT DEFINITION table used for defining a dynamic table.
+
+| key                     = | P4RT:DEFINITION:table_name | ; Dynamic table name |
+|---------------------------|----------------------------|----------------------|
+
+
+
+```
+127.0.0.1:6379> keys P4RT*
+1) "P4RT:DEFINITION:ACL_ACL_PRE_INGRESS_TABLE"
```

Comment:
Did we consider separate DB instance e.g P4RT_DB for P4 table definitions? this would help to avoid any accidental flush of the P4RT configs in APPL_DB during warm restart, swss restart..etc.

Reply:
P4RT tables share very similar requirements during warm restart, swss restart etc. In some cases, local applications require more time to learn the new conditions following a restart. However, the remote controller is not affected by a local restart and can immediately push its intent as soon as the local restart is complete. There is more similarity than differences and the longer term approach is merging with the local tables.